src/plonk/circuit.rs

```diff
@@ -1072,11 +1156,280 @@ impl<F: Field> ConstraintSystem<F> {
+                // is multiplying a value that is constrained to equal zero.
+                let mut expr = query.clone();
+                let mut root = F::one();
+                for _ in 0..combination.len() {
+                    if root != assignment {
+                        expr = expr * (query.clone() - Expression::Constant(root));
+                    }
+                    root += F::one();
+                }
+
+                assert!(selector_replacements[selector].is_none());
+                selector_replacements[selector] = Some(expr);
+
+                assignment += F::one();
+            }
+        }
+
+        // All selectors have a combination now.
+        for selector in selector_map {
+            self.selector_map.push(new_columns[selector].unwrap());
+        }
+        let selector_replacements = selector_replacements
+            .into_iter()
+            .map(|a| a.unwrap())
+            .collect::<Vec<_>>();
+
+        fn replace_selectors<F: Field>(
+            expr: &mut Expression<F>,
+            selector_map: &[Expression<F>],
+            must_be_nonsimple: bool,
+        ) {
+            *expr = expr.evaluate(
+                &|constant| Expression::Constant(constant),
+                &|selector| {
+                    if must_be_nonsimple {
+                        // Complex selectors are prohibited from appearing in
+                        // expressions in the lookup argument by
+                        // `ConstraintSystem`.
+                        assert!(!selector.is_simple());
+                    }
+
+                    selector_map[selector.0].clone()
+                },
+                &|query_index, column_index, rotation| Expression::Fixed {
+                    query_index,
+                    column_index,
+                    rotation,
+                },
+                &|query_index, column_index, rotation| Expression::Advice {
+                    query_index,
+                    column_index,
+                    rotation,
+                },
+                &|query_index, column_index, rotation| Expression::Instance {
+                    query_index,
+                    column_index,
+                    rotation,
+                },
+                &|a, b| a + b,
+                &|a, b| a * b,
+                &|a, f| a * f,
+            );
+        }
+
+        // Substitute selectors for the real fixed columns in all gates
+        for expr in self.gates.iter_mut().flat_map(|gate| gate.polys.iter_mut()) {
+            replace_selectors(expr, &selector_replacements, false);
+        }
+
+        // Substitute non-simple selectors for the real fixed columns in all
+        // lookup expressions
+        for expr in self.lookups.iter_mut().flat_map(|lookup| {
+            lookup
+                .input_expressions
+                .iter_mut()
+                .chain(lookup.table_expressions.iter_mut())
```

Comment:
At https://github.com/zcash/halo2/pull/347/files#diff-c7d3f048cff35b9298e74eb82f24b7755a72927c699f387552c0ac97f7c6c4a8R993 we required tables to be tuples of `Column<Fixed>`, so it shouldn't be necessary to substitute table expressions. (But it looks as though it should work anyway, so what was the rationale for the `Column<Fixed>` restriction?)

Reply:
I wanted to prevent selectors from appearing in the lookup tables. The quick and dirty solution is to just restrict it to `Column<Fixed>` on the forward facing API (the `lookup` method on `ConstraintSystem`) and not have to go through the entire lookup argument and modify it to special case the tables. This way we can reintroduce expressions for tables if we need them someday.